libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -789,67 +799,436 @@ bool GCS_MAVLINK::handle_mission_item(mavlink_message_t *msg, AP_Mission &missio
+
+    case MAVLINK_MSG_ID_POSITION_TARGET_GLOBAL_INT:
+        return MSG_POSITION_TARGET_GLOBAL_INT;
+
+    case MAVLINK_MSG_ID_ADSB_VEHICLE:
+        return MSG_ADSB_VEHICLE;
+
+    case MAVLINK_MSG_ID_BATTERY_STATUS:
+        return MSG_BATTERY_STATUS;
+
+    case MAVLINK_MSG_ID_AOA_SSA:
+        return MSG_AOA_SSA;
+
+    case MAVLINK_MSG_ID_DEEPSTALL:
+        return MSG_LANDING;
+
+// can't really stream out named float messages!
+    // case MAVLINK_MSG_ID_XYZZY:
+    //     return MSG_NAMED_FLOAT;
+
+    default:
+        return MSG_LAST;
+    }
+    return MSG_LAST;
+}
+
+bool GCS_MAVLINK::set_mavlink_message_id_interval(const uint32_t mavlink_id,
+                                                  const uint32_t interval_ms)
+{
+    const ap_message id = mavlink_id_to_ap_message_id(mavlink_id);
+    if (id == MSG_LAST) {
+        gcs().send_text(MAV_SEVERITY_INFO, "No ap_message for mavlink id (%u)", mavlink_id);
+        return false;
+    }
+    return set_ap_message_interval(id, interval_ms);
+}
+
+bool GCS_MAVLINK::send_message_in_delay_callback(const ap_message id) const
+{
+    // No ID we return true for may take more than a few hundred
+    // microseconds to return!
+
+    if (id == MSG_HEARTBEAT) {
+        return true;
+    }
+
+    if (in_hil_mode()) {
+        // in HIL we need to keep sending servo values to ensure
+        // the simulator doesn't pause, otherwise our sensor
+        // calibration could stall
+        if (id == MSG_SERVO_OUT ||
+            id == MSG_SERVO_OUTPUT_RAW) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+// linearly scan the message queue to find the next message to send out:
+void GCS_MAVLINK::point_next_deferred_message_to_next_message_to_send()
+{
+    // assume failure:
+    _next_deferred_message_to_send = no_deferred_message_to_send;
+
+    bool first_done = false;
+    uint32_t winning_time = 0;
+    bool repeating_message_found = false; // so we can mark chan as streaming
+
+    for (uint8_t i=0; i<ARRAY_SIZE(deferred_messages); i++) {
+        deferred_message_t &e = deferred_messages[i];
+        if (e.id == MSG_LAST) {
+            continue;
+        }
+        if (e.id == MSG_NEXT_PARAM && e.interval_ms == 0) {
+            // force parameters to *always* get streamed so a vehicle
+            // is recoverable from bad configuration:
+            e.interval_ms = 100;
+        }
+        if (e.interval_ms) {
+            repeating_message_found = true;
         }
-        next_deferred_message++;
-        if (next_deferred_message == ARRAY_SIZE(deferred_messages)) {
-            next_deferred_message = 0;
+        if (e.send_not_before_ms == 0) {
+            continue;
+        }
+        if (first_done &&
+            e.send_not_before_ms > winning_time) {
+            continue;
         }
-        num_deferred_messages--;
+        if (hal.scheduler->in_delay_callback() &&
+            !send_message_in_delay_callback(e.id)) {
```

Comment:
Currently we do a one-to-one mapping.  But that's already wasteful as things like named-message-floats will never take up a slot.  As we start adding more and more messages that people *might* want, we don't necessarily want to increase the size of this array to match.  In fact, I'm hoping in the future we might trim messages we currently automatically send.

Reply:
I don't get what you're saying with "we might send a message in delay callback that we shouldn't"...

Reply:
"_next_deferred_message_to_send that isn't actually the one that needed to be sent next" -> yeah, I can see that now.  The only fix I can see for that is to have another variable indicating whether the _next_deferred_message_to_send was calculated while within a delay callback :-(

Reply:
> But that's already wasteful as things like named-message-floats will never take up a slot.

It will take a slot every time it's sent right? You can't set the rate from a GCS, but the message is still sent as an one-time only message.

> As we start adding more and more messages that people might want, we don't necessarily want to increase the size of this array to match. In fact, I'm hoping in the future we might trim messages we currently automatically send.

OK, I would decouple it now then, even if the number is the same for now.